### PR TITLE
qa/tasks/cephadm: add generic templating where subst_vip was used

### PR DESCRIPTION
**Pull request by phlogistonjohn**
After some discussion with @adk3798 I decided to add more generic templating functions to qa/tasks/cephadm.py. Initially, this simply replaces the simple but rigid string-replacement based "templates" like `{{VIP0}}` with generic templates based on jinja2.

AFAICT, jinja2 is already a dependency of teuthology so this PR doesn't add a new dependency.

Later this work will be reused when I add support for deploying a Samba AD DC for testing cephadm's smb service. It should also serve as foundational work for other cases where we have a value that varies from test to test and need to express that in a service spec or such.

NOTE: this only touches the tasks python code. It should primarily be tested by ensuring there are no regressions in a teuthology orch suite run.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional) - sort of
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), <s>opened tracker ticket</s> tests only
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests
  - [x] Updates teuthology test infra code only.

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2024-02-27_05:08:20-orch:cephadm-wip-adk-testing-2024-02-26-1600-distro-default-smithi/

not going to go over the failures in the regular detail as we're still cleaning up the `in cluster log` stuff, but it was mostly just things we still need to add to ignorelists and the mds_upgrade_sequence test having issues. Only test that failed due to PRs in the run was https://pulpito.ceph.com/adking-2024-02-27_05:08:20-orch:cephadm-wip-adk-testing-2024-02-26-1600-distro-default-smithi/7574464 so we can't merge the PRs making changes to mgr/nfs but everything else seems to be okay.

File: qa/tasks/cephadm.py [hunk 1]
```diff
@@ -27,13 +30,69 @@
...
-from tasks.vip import subst_vip
 
 CEPH_ROLE_TYPES = ['mon', 'mgr', 'osd', 'mds', 'rgw', 'prometheus']
 
 log = logging.getLogger(__name__)
 
 
+def _convert_strs_in(o, conv):
+    """A function to walk the contents of a dict/list and recurisvely apply
+    a convertion function (`conv`) to the strings within.
```
Comment (adk3798) [inline qa/tasks/cephadm.py:41]: NIT:
```suggestion
    a conversion function (`conv`) to the strings within.
```

File: qa/tasks/cephadm.py [hunk 2]
```diff
@@ -27,13 +30,69 @@
...
+    string values without the risk of changing the structure of the yaml/json.
+    """
+    jenv = getattr(ctx, '_jinja_env', None)
+    if jenv is None:
+        loader = jinja2.BaseLoader()
+        jenv = jinja2.Environment(loader=loader)
+        setattr(ctx, '_jinja_env', jenv)
+    rctx = dict(ctx=ctx, config=config, cluster_name=config.get('cluster', ''))
+    _vip_vars(rctx)
+    log.warning(f"{rctx=}")
```
Comment (adk3798) [inline qa/tasks/cephadm.py:79]: do we want this logging at warning level? It seems to just always be doing this rather than in an error case.

**Review by adk3798 — APPROVED** [review]
Worked when testing, e.g. https://pulpito.ceph.com/adking-2024-02-27_05:08:20-orch:cephadm-wip-adk-testing-2024-02-26-1600-distro-default-smithi/7574467/ used the vip and vip.exec tasks and passed.

File: qa/tasks/cephadm.py (hunk 1 above)
Comment (phlogistonjohn) [reply qa/tasks/cephadm.py]: Fixed.

File: qa/tasks/cephadm.py (hunk 2 above)
Comment (phlogistonjohn) [reply qa/tasks/cephadm.py]: Removed.

**Pull request merged**